draft/0004-user-sign-in-experience.md

```diff
@@ -0,0 +1,669 @@
+
+We will need to keep two sets of APIs in sync to ensure that the experience APIs are up-to-date with the internal interaction APIs. This will require additional development and testing effort to maintain the consistency between the two sets of APIs.
+
+### 6.4 Performance overhead
+
+To maintain interaction status and decouple API logic, multiple API requests are required from the client side. For example, an additional submit request must be called each time to submit and validate the interaction requirements, even for a simple username and password sign-in flow. Instead of a single sign-in request, an extra submit request is necessary. This increases the number of API calls, potentially leading to higher latency and reduced performance.
+
+## 7. Rationale and alternatives
+
+Logto provides a variety of authentication and verification methods to meet the diverse needs of developers and end-users. We have been constantly improving the sign-in experience configurations and providing extra customization options for the developers to build a flexible and secure sign-in experience.
+As the sign-in experience is a critical part of the user authentication process, and it is hard to enumerate all the possible uses cases, we believe that opening up the experience APIs to the developers with a customizable front-end user interface will provide more flexibility and control over the sign-in experience.
+
+### 7.1 Considerations
+
+1. Developer Friendly: The current internal interaction APIs are not well-documented and are difficult for developers to use. These APIs are tightly coupled with Logto's server-side interaction logic and involve numerous Logto-specific concepts and definitions, resulting in a steep learning curve. We believe that providing a new set of well-documented and user-friendly experience APIs will significantly ease the development process for creating customized sign-in experiences. Additionally, we need to offer a detailed explanation of the underlying user interaction structure to help developers better understand the sign-in experience.
+
+2. High flexibility: We have identified several limitations in the current internal interaction APIs, such as the lack of support for custom user profile fields and the increasingly bloated interaction structure within the Logto server. By introducing the new experience API, we can redefine the interaction in Logto, enhancing the flexibility and sustainability of the sign-in experience.
+
+### 7.2 Alternatives
+
+Considering no extra new features or customization options are added, instead of introducing a new set of experience APIs, we could consider providing comprehensive documentation for the existing internal interaction APIs. This approach would minimize the risk of breaking changes to the sign-in experience. However, even with detailed documentation, developers may still find it challenging to understand and effectively use these internal interaction APIs.
```

Comment:
```suggestion
Considering no extra new features or customization options are added, instead of introducing a new set of APIs, we could consider providing comprehensive documentation for the existing internal interaction APIs. This approach would minimize the risk of breaking changes to the sign-in experience. However, even with detailed documentation, developers may still find it challenging to understand and effectively use these internal interaction APIs.
```